format/FlightSql.proto

```diff
@@ -1842,6 +1842,94 @@ message ActionCancelQueryResult {
   CancelResult result = 1;
 }
 
+/*
+ * Request message for the "Close Session" action.
+ */
+message ActionCloseSessionRequest {
```

Comment:
Is the assumption for all of these that the session/client identifier will be transmitted out-of-band (via headers, etc.)?

Reply:
Correct, and to address (will comment "see above") this and other comments the intention is that the Session-Id will come back via Set-Cookie which will allow support for session persistence by simply reusing standard cookie middleware/no additional code.  Obviously the client can if it wishes do something fancier to share them, and the server must implement either (one of) SetSessionOptions or Resource-Path request header handling in order to support this feature.

Clients will require try-SetSessionOptions-and-fallback-to-headers logic for full support.  I looked at rolling this into the FlightSqlClient constructor, and while it was feasible to make that non-breaking the snag was that introducing by-default middleware injection could easily break a client app's existing middleware scheme unless we at least introduced some (unreliable, so no) detection logic to see if any ClientCookieMiddlewareFactory was already provided by the app.

Reply:
I don't think I quite follow this as an answer; my concern here is that implementing sessions seems to require a lot of magic application code to handle all the different code paths in the server, when it seems we should be providing a standard interface here that gets turned into whatever underlying combination of headers and RPC requests. For instance, how does the server get the session identifier _at all_ right now? You hard-code a function that looks for a session cookie and then falls back to looking for headers?

Reply:
I'll try this a little differently as I know I've been taking about nitty-gritty specifics and maybe missing some big-picture coherence.

There is definitely no magic client app code required for this, it becomes a single call for one or more session parameters unless one wants to update them.

Server-side, we were on the fence (mostly considering previously-expressed concerns about more Arrow code to maintain) but if you don't mind the bit of extra helper module code we can definitely implement and provide a baseline server middleware for this.  Fundamentally the server generates the session identifier initially (on the first SetSessionOptions call, presumably sent without any session ID cookie header.

The header functionality, after some earlier redesign, has ended up being something that we're really only looking at IF an application developer needs to have some equivalent functionality baked into an app they've locked to a legacy version of Arrow for whatever reason.  Otherwise there is no reason for the header fallback under any circumstances with up-to-date Arrow versions.  Therefore this is something that's just going to be suggested as a legacy implementation option for use with stale Arrow versions.

On to the bigger picture, the actual operation flow is roughly like so—hopefully this provides some clarity:

* Client is initialized—with ClientCookieMiddleware or a custom alternative—and connects to server
* Client issues a SetSessionOptions DoAction with whatever session parameters, and no session identifier cookie header
* Server handler queries the call-local middleware instance which is aware that no session identifier was specified, and therefore returns a newly-initlaized session identifier.
* Server handler pushes the options down into session middleware, and assuming that is successful then returns success status(es) for the applicable options being set, along with the Set-Cookie header containing the session identifier
* The Flight SQL session proceeds as normal, with the ClientCookieMiddleware handling the session persistence.

Reply:
Ok, thanks - that's what I was trying to get at.

Without the middleware, how do you expect servers to do all this? 

What is the concern about having middleware? Is it possible to build a set of low-level helpers (presumably, what a server would need to do if it didn't have middleware) as well as a higher-level middleware using those helpers?
 
Also, I am not sure the headers are only a legacy path. Not all clients are willing or able to maintain persistent state (and/or persistent state may be unreliable). So we should consider the headers as a first-class path as well.

Reply:
The concern was simply minimizing how much more stuff we're jamming into the Arrow package (and if it's not really necessary etc), however I think this is both important enough (to clarify the flow described above so we're not giving users/developers a puzzle to put together) and not more bulk added to the core code (Middleware being an optional helper) so it's probably worth adding an interface and impl to the package.  Do the guidelines being used discourage MI?—by default I'm looking at implementing the core Middleware interface plus the session object getter via another interface that the new server session middleware would also inherit.

Reply:
Re: the lower-level helper question, I think the middleware implementation we were debating whether to include is (a) at the right interface level (bypassing the middleware interface probably doesn't make a lot of sense, and it's necessary to use the Flight Server Middleware interface to allow the handlers to juggle the inbound and particularly outbound session ID cookie local to the call, and (b) a complete reference as to the required mechanics to make everything work should anyone feel the need to try to do the same thing differently.

Re: header support, I'm not 100% sure if you're talking about the session ID or the options themselves.  I am reticent to provide default (part of Arrow vs. app-defined) support of arbitrary option headers, but as these are explicitly not part of the session functionality and work via headers that the client handlers can already access and handle I think supporting that is a no-op.  As to providing headers (instead of full client middleware) as a fallback for keeping/sharing the session ID across e.g. distributed clients or whatever, the session ID is already a header (albeit a cookie) so I can simply provide the accessor we were considering for getting the session ID out of the provided middleware implementation (probably a subclass with just the extra accessor(s)) and then client apps can do as they please with that data.

Reply:
ServerSessionMiddleware (builds but I am still getting tests sorted) has been added to the PR as discussed.  Please let me know if you have any thoughts on ways you'd prefer the mechanics to be different.  (ServerSessionMiddleware also needs a bit more documentation possibly, again comments welcome but I will be reviewing that anyways.)

Reply:
> I'm referring to the options themselves; I thought the original proposal was that servers should also support sending options as arbitrary headers, since the existing JDBC driver does that.

At least internally we're looking at moving a bunch of things off of being sent as headers and over to session options.  For backwards compatibility internally as well as in the wild I'm inclined to pursue making the JDBC and ODBC drivers detect the Session Option doAction verbs and use them if present and only fall back to headers if the verbs are absent.  (Relatedly, it might be useful to have the doAction listing populated dynamically or overridden by the app's client subclass as the discrepancy between the listing and what Actions do and do not return a not implemented status message is a bit confusing for this and other purposes...)

Reply:
To circle back here, expectations should be clearly documented in the Protobuf. So here, it should be explicitly stated that the client and server are using a shared header to store the session state, and that the header should be the HTTP Cookie/Set-Cookie header.